zio_compress: introduce max_size threshold

gmelikov opened the pull request:
Created brand new PR because of some problems with rebasing old one #9311, + updated description.
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
87.5% minimum compressratio is way to large, on 1TB with recordsize 16M such artifical limitation may lead to up to 125GB of wasted space in worst case (but you must be very, very unlucky). 

So i propose to check only for max ashift of pool, it may save you nearly 99% of compressratio on large recordsizes.

And I think if you enable slow compression on dataset you will prefer better compressratio over decompression absence for data nowadays.

So, pros:
- better compressratio
- better IO if decompress is faster than disks
- we're already compressing everything
- it may be tuned for not so compressible data via module parameter

Cons: 
- if you have data, which used to be written uncompressible due to previous code - you will see better compressratio at the price of more CPU load

### Description
<!--- Describe your changes in detail -->
Now default compression is lz4, which can stop
compression process by itself on incompressible data.
So we can check only for common sector size.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
Generate 12% (just as needed to get into 87.5% threshold) compressible file with fio:
```
# cat fio.conf
[workload]
bs=128k
ioengine=libaio
iodepth=1
numjobs=1
direct=1
runtime=10
size=100m
filename=bigfile
rw=read
thread
unified_rw_reporting=1
group_reporting=1
buffer_compress_percentage=12
refill_buffers
buffer_pattern=0xdeadbeef
```
```
zfs set compression=lz4 rpool/test
```

version | recordsize | ashift | written | diff (from original)
--- | --- | --- | --- | ---
0.7.13 | 1M | 12  | 100M 
master_patched  | 1M | 12  |  88.8M   | ~11.2%

So there may be up to ~12.5% compression gain for hardly compressible files.

Tested manually in VM, didn't run actual system on this code yet.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

Ornias1993 commented:
@gmelikov Thanks for pointing me here.
I can at least confirm that 12% compressible writes still don't get compressed as today on MASTER, so this is not just an issue with 0.7.13.

Good work on this, LGTM, this PR deserves some more love! 👍

ahrens commented on module/zfs/zio_compress.c:
```diff
@@ -120,8 +120,23 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len)
+		 * Write compressed only if there is at least
+		 * one sector compressed
+		 */
+		d_len = s_len - MIN(s_len, zfs_compress_threshold_bytes);
```
Given the check above, `s_len` can't be less than `zfs_compress_threshold_bytes` here.
```suggestion
		d_len = s_len - zfs_compress_threshold_bytes;
```

ahrens commented:
I would prefer to eliminate the module parameter, and instead "use sector size **of this pool** as threshold".  I.e. always make the threshold 1 sector (spa_max_ashift).

gmelikov commented:
@ahrens hmm, I agree with you, we don't need to use larger threshold here with lz4 and future zstd.

Will get back to this PR in a week.

gmelikov commented on module/zfs/zio_compress.c:
```diff
@@ -120,8 +119,12 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len)
+		return (s_len);
+
+	/* Write compressed only if there is at least one sector compressed */
+	d_len = s_len - MIN(s_len, zfs_compress_threshold_bytes);
```
Fixed, now we're using pool's max ashift.

gmelikov commented:
Updated,
- I've removed module parameter and use max_ashift of pool;
- tested interoperability with old code, manually imported pool with new code, created data with 89% effective compression, imported it with old code, scrub/read/write is ok, rewrote it, imported with new code, scrub/read/write ok;
- nightly ztest is almost all green, I've got just one `ASSERT at ztest.c:6050:ztest_scrub()/home/debian/zfs/cmd/ztest/.libs/ztest(+0x9444)[0x564c51a51444]`, but looks like it's not related to this patch
- I've found only one place in code where we recompress data to checksum it - https://github.com/openzfs/zfs/pull/9416/files#diff-163982413ed6b6861e08633c89a77395R1759 , so looks like encrypted dataset and ARC without compression may trigger a checksum error on pool with 89+% compressed data on old code, @tcaputi could you please help me here and prove it or give some information. I can't reproduce this case, looks like I'm wrong;
- I looked through all `zio_compress_data()` usages and can conclude that if data wasn't compressed - it will be saved as `ZIO_COMPRESS_OFF`, so if we really need to recompress data without ashift info and check anything - we can just compress it with zero threshold;
- Old code (unless of uncompressed ARC and encryption case) doesn't recompress data for checksums, so if that case will be resolved - we can use this patch without feature flag!

tcaputi commented:
@gmelikov without looking at it too closely i think i might know whats going on. I think there are 2 things as play which are combining to result in the ARC's "recompress before checksumming" to work ok, but I think it is a pretty brittle dependency.

For some quick background, we need to store data in the L2ARC exactly the same as it is on disk in order for the decryption to work. This is why we do the re-compression here in the ARC layer.

When the data gets recompressed, the ARC uses the compression algorithm that it stored in the ARC header. This info comes from the bp when the block first enters the ARC. This patch doesn't change the way compression is done (AFAICT), it just changes the threshold at which we decide the compression is "worth it". However, no matter what that threshold is, if ZFS decides not to compress the block it sets the compression algorithm to `ZIO_COMPRESS_OFF`. So the encryption code will only recompress it if it was actually compressed. This code does NOT have the same check for whether or not the compression was worth it. It simply relies on the value from the bp, so it ends up doing the right thing in all these cases.

That being said, this discussion is making me wonder if QAT compression support breaks L2ARC with encryption + no compression. I suspect it does, but that is a very niche bug with very minor repercussions in that case (the data simply fails to read from the L2ARC) and checks the main pool instead. Probably a problem for another time.

Anyway, hope that helps.

gmelikov commented:
@tcaputi Thank you for insight, you're right, if it was compressed - we can use zero threshold here too, so I was wrong here (how glad I am)!

So, this patch is ready to review, I've rebased it.

gmelikov commented on module/zfs/arc.c:
```diff
@@ -1755,7 +1756,7 @@ arc_hdr_authenticate(arc_buf_hdr_t *hdr, spa_t *spa, uint64_t dsobj)
 		abd_take_ownership_of_buf(abd, B_TRUE);
 
 		csize = zio_compress_data(HDR_GET_COMPRESS(hdr),
-		    hdr->b_l1hdr.b_pabd, tmpbuf, lsize);
```
Nope, steps to get problem here:
- Use code from this PR
- get 89+% compressed encrypted block
- L2ARC attached to pool
- Load pool with pre-PR code
- `zio_compress_data` will not compress this block again here due to old 12.5% threshold inside it
  
Heck, because it's the problem - I'll add a read-only feature flag only to make it safe for this case. In future `zio_compress_data` will have threshold argument so we won't get in this probem again.

ahrens replied:
I see, if you move the pool to a system with older software, it doesn't correctly write to the L2ARC, because even though the block was compressed, the old L2ARC-writing code will not compress it if it doesn't save at least 1/8th of the space.  This only applies if the old system has disabled compressed ARC, right?

Introducing a readonly-compatible feature flag is a bit unfortunate.  It would mean maintaining the old 1/8th code forever.  I wonder how common it is to disable compressed ARC, nowadays?  It might be reasonable to say that if you disable compressed ARC, and use L2ARC, and bring a newer pool back to old bits, then some of the data won't be able to "stick" in the L2ARC.  I mean, we previously considered (but didn't implement) saying that disabling compressed ARC makes L2ARC not work with compressed blocks at all (i.e. removing the recompress-on-L2ARC-write code entirely).

gmelikov replied:
@ahrens yes, you've got the point. The funny thing, that I can't reproduce problems in such theoretic scenario, l2arc works well under manual tests with data from new code, but with old module and disabled ARC compression. 

There is a "best effort" thing near `arc_hdr_authenticate` function, maybe it's the case?

I've tried to add a feature flag, and code began to transform from beautiful and easy to read info ugly if-if monster. I don't want to do this :(

If it's already a problematic and rare case (compressed l2arc with encryption on old code with pool and data from new code), *and* we are ready to go with this - I propose to merge it as is without a feature flag, PR is ready for it. I can add this comment somewhere (I'd be glad if someone gives me exact place for this).

adamdmoss commented:
(as an aside, the comments here talk about a >1MB recordsize a few times - I thought 1MB was the upper limit?)

gmelikov commented:
@adamdmoss by default it is, but technically 16MB is an upper limit now, you may just change this module parameter and voila https://github.com/openzfs/zfs/wiki/ZFS-on-Linux-Module-Parameters#zfs_max_recordsize

It's good if you have really large block workload with rare read.

gmelikov commented:
> its still questionable if this correct then because your threshold is then some sort of alignment per sector and you are using this shift as threshole in zio_write_compress too which is unrelated to your l2arc case. a standard record size oriented block has a different alignment. so just using a check like destsize<srcsize should be enough from my point of view

@BrainSlayer I don't agree, sector size is used for threshold because we can save space on disk only if we saved at least one sector size of space, not less.

BrainSlayer commented:
but a recordsize of 1M (or 128kb by default) will span over multiple sectors. so doing it on sector alignment is okay, but sectorsize as maximum boundary makes no sense.

gmelikov commented:
@BrainSlayer can't uderstand what you mean, 
`d_len = s_len - compress_threshold;` here for 128K block and 4K sector size we'll set it to "compress result must be less than 120K or write it as is", you can see it here https://github.com/openzfs/zfs/pull/9416/files#diff-cddcbcd27b593866d4d3a2ffbc06ff4bR134

Besides, you can add comment to exact line of code to be more precise.

BrainSlayer commented:
from my understanding a physical sectorsize is much smaller than a block size. so i would solve this in the following way targetsize = inputsize - (inputsize % sectorsize). so we ensure that its sector aligned, no matter how big the blocksize is

BrainSlayer commented:
but a uncompressed block will still ignore this rule of course. so the question is if it really makes any benefit if we align it, because if its uncompressed it will be still cross the limit

BrainSlayer commented:
regarding your formula. 128kb - 4K = 124K but 128K is sector aligned already. so no need to subtract anything

thulle commented:
@BrainSlayer It has nothing to do with alignment though? It's a threshold to see if the compression actually saves space? In that scenario if it's more than 124KB we're writing 16 4KB sectors anyway and can just write it uncompressed, if it's less we've compressed enough to save a sector.
The original code does the same calculation but with a higher threshold: 
```
/* Compress at least 12.5% */
	d_len = s_len - (s_len >> 3);
```

GregorKopka commented:
@BrainSlayer ZFS stores data in 2^ashift sized blocks on-disk, this is distinct from both the sector size the drives expose on their external interface (SAS, SATA, whatever) and the volblock-/recordsize configured for the dataset. Storing data compressed only makes sense when it needs _less_ on-disk (ashift) blocks than the uncompressed data would, as the overhead of having to decompress data on access is then at least offset by a reduced amount of data that needs to be read from the pool.

Apart from that: this isn't IRC or discord, please try to write a comment in one instead of multiple, as each comment spawns a mail toward each participant of this issue.

BrainSlayer commented:
@thulle yes, but the 12.5% is a wild value. i mean after compression the ratio is detected, but the performance for doing the compression is already lost. so even if its a small ratio, it still makes sense. with algorithms like gzip with lower decompression speed such a ratio makes sense, but high performance algorithms like LZ4 or ZSTD doesnt need such a hard ratio.

BrainSlayer commented:
@GregorKopka yes. but a simply compressedsize < sourcesize match would be already enough. we dont need to subtract another sector size as ratio. in our example with 4K sector size the ratio limit is 124K if just 123K is stored 1K is wasted. but that would be the same if we dont substract it and just 127K is stored. we just limit the efficency. from the performance itself it wouldnt make any different since algorithms like LZ4 or ZSTD are faster than physical drive access in most usecases

thulle commented:
@BrainSlayer Which is the reason for this pull request? To get rid of that 25% "wild value".

And if we use compressedsize < sourcesize than a 128K record compressed to 127K would still use 128K on disk and we'll just waste some minor time decompressing again when reading. Can just as well save it uncompressed for slightly better performance.

BrainSlayer commented:
@thulle i cannot answer this question. read the topic of this pull request. its not mine

yes 127kb will still waste 1kb. but its the same waste if its stored uncompressed and will not make any difference at reading. but if we store just less than 124kb the propability of storing 128k blocks uncompressed is higher which also doesnt make any sense. and sector sizes can also be very different. i mean ssd's have typical 512 byte physical addressing size no matter how the filesystem is configured

thulle commented:
@BrainSlayer With the caveat that I'm not familiar with the code but wanting to spare the actual devs from explaining what I perceive as misunderstandings on your end:

It does make a difference. Comparing reading 128KB vs reading 128KB and running it through an decompression algorithm, the pass through decompression - however fast it is - would just be wasted CPU-cycles. Thus it does makes sense to only write things compressed if we actually save any diskspace by doing so. And since we have to write things in ashift-sized blocks this pull request changes the threshold from that arbitrary 25% compression ratio to "does this compression save at least one ashift-sized block of space?" To accomplish this we compare to a threshold that's the equivalent of if (number of ashift-sized blocks compressed) < (number of ashift-sized blocks uncompressed).
If ashift is 12 an 128KB record has to compress to 124KB (128KB - one single saved 4KB block) to be saved compressed. With ashift 14 the threshold becomes 112KB (128KB - one single saved 16KB block).

BrainSlayer commented:
i agree that one sector size at least would make sense for minimum compression threshold, since anything more would not greate any benefit for physical read/write efforts. but consider that these days sector sizes are more or less virtual since the introduction of ssd's which will become more and more important in future. and the decompression for most modern algorithms are much faster than reading any data from storage.  so the question is if it creates any performance benefit to go this way and making the code more complex.

gmelikov commented:
@BrainSlayer I've got your point and generally agree with it, but I think that one simple check in this exact case is still faster than decompression of block without any space gains. More on that - we still want to check if compression didn't make it even worse than original size, so this check is nearly free already, IMHO.

ahrens commented on man/man8/zfsprops.8:
```diff
@@ -855,11 +855,11 @@ is selected, compression will explicitly check for blocks consisting of only
-and the block saved uncompressed. Note that when the logical block is less than
-8 times the disk sector size this effectively reduces the necessary compression
-ratio; for example 8k blocks on disks with 4k disk sectors must compress to 1/2
+Block must be compressed by at least max ashift of pool vdevs, otherwise
+the compression will not be considered worthwhile and the block saved
+uncompressed. Note that when the logical block is less than 8 times the disk
+sector size this effectively reduces the necessary compression ratio;
+for example 8k blocks on disks with 4k disk sectors must compress to 1/2
 or less of their original size.
```
I took a stab at re-explaining these concepts, although I feel we could also just drop this paragraph, as it is fairly deep in the implementation details.

```suggestion
All blocks are allocated as a whole number of sectors (chunks of 2^ashift bytes, e.g. 512B or 4KB).  Compression may result in a non-sector-aligned size, which will be rounded up to a whole number of sectors.  If compression saves less than one whole sector, the block will be stored uncompressed.  Therefore, blocks whose logical size is a small number of sectors will experience less compression ratio  (e.g. for recordsize=16K with 4K sectors, which have 4 sectors per block, compression needs to save at least 25% to actually save space on disk).
```

ahrens commented on module/zfs/zio_compress.c:
```diff
@@ -102,7 +98,8 @@ zio_compress_zeroed_cb(void *data, size_t len, void *private)
 }
 
 size_t
```
Let's add a comment here explaining the `compress_threshold` parameter, e.g.:
`Only compress if it results in saving at least compress_threshold bytes.`

ahrens commented on module/zfs/zio_compress.c:
```diff
@@ -120,8 +117,24 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len)
+	 * we can skip these checks for reproducible results,
+	 * so set compress_threshold=0 for reproducibility of ARC checks.
+	 */
+	if (compress_threshold > 0) {
```
I don't think we need the `if` statement.  If `compress_threshold` is zero, the code in this block will still do what we want.

ahrens commented on module/zfs/arc.c:
```diff
@@ -8531,7 +8532,8 @@ l2arc_apply_transforms(spa_t *spa, arc_buf_hdr_t *hdr, uint64_t asize,
 		tmp = abd_borrow_buf(cabd, asize);
 
-		psize = zio_compress_data(compress, to_write, tmp, size);
+		psize = zio_compress_data(compress, to_write, tmp, size,
+		    1 << spa->spa_max_ashift);
```
This case seems similar to arc_hdr_authenticate(), in that we know the data was previously compressed, and we are trying to recreate the same on-disk bits that we had before.  So I think we should pass `compress_threshold=0` here.

ahrens commented on module/zfs/arc.c:
```diff
@@ -9899,7 +9901,8 @@ l2arc_log_blk_commit(l2arc_dev_t *dev, zio_t *pio, l2arc_write_callback_t *cb)
 	psize = zio_compress_data(ZIO_COMPRESS_LZ4,
-	    abd_buf->abd, tmpbuf, sizeof (*lb));
+	    abd_buf->abd, tmpbuf, sizeof (*lb),
+	    1 << dev->l2ad_spa->spa_max_ashift);
```
I don't think it really matters what threshold we use for the l2arc log block, in that the impact either way will be trivial to nonexistant.  But: I think it would make more sense to use dev->l2ad_vdev->vdev_ashift here, since we know exactly which device we're writing to, and IIRC spa_max_ashift doesn't take into account L2ARC devices.

BrainSlayer commented:
> @BrainSlayer I've got your point and generally agree with it, but I think that one simple check in this exact case is still faster than decompression of block without any space gains. More on that - we still want to check if compression didn't make it even worse than original size, so this check is nearly free already, IMHO.

you dont need to check if its getting worse. the check < sourcesize is already catched. if compression result is bigger or equal source size, the block will be stored uncompressed. this case is already handled. no need todo anything

BrainSlayer commented:
sorry. i have to correct my sayings. i handled this case in my zstd PR code, but its not handled by the upstream code. so the zstd PR does handle the case that the result is bigger than the source for all available compression algorithms

gmelikov commented:
@ahrens thanks, agree with everything.

@BrainSlayer we should deal with all compression types here, unfortunately, but we may reconsider this logic in future without additional feature flags.

BrainSlayer commented:
@gmelikov i just think a compression threshold limiter parameter is a logic consequence in this development here. why do we decide what a user wants? give him the choice. black box concepts are never good. we can make sane default values like threshold 0 = auto. (your way) or any other value = percentage budged from source size.

adamdmoss commented:
Personally I'd like the threshold to be tunable, perhaps with 15% retained at the default.

For write-once-read-never data and other situations dominated by write speed or where space is critical, using a ~1% threshold makes a lot of sense.

For the out-of-the-box case of lz4 where compression is an optimization and space-saving is merely a great side-effect, a larger threshold (say 15%) feels right for keeping us off the _relatively_ slow-read path.

Ornias1993 commented:
@gmelikov Is there any update/progress on this?

gmelikov commented:
@Ornias1993 unfortunately, there is still a minor case with l2arc and compatibility with older code, I'll try to work on it on holidays, there are some more important things with CI in my work-in-progress list now.

adamdmoss commented:
FWIW I added the ability to set this % as a per-dataset value while I was working on some other stuff, but ran into similar subtle issues where recompressed MAC was expected to match original compressed MAC but (AFAICT) the result implicitly depends on the discussed cutoff threshold remaining constant when compressed ARC is disabled.

gmelikov commented:
@adamdmoss if you already implemented such threshold as dataset param (which I like too), then I propose to close this PR in favor of your's future one, I'd be glad to review it. In addition, this PR as is has problems with dbufstat tests.

I'll drop 12.5% general threshold removal idea then, and look into basic logic about other conditions, for example I've got to `embedded_data` feature, and even with 12.5% threshold we'll affect it. So we can't just ignore compressions for blocks smaller than ashift size, we should check size after compression for `embedded_data` feature.

adamdmoss commented:
I dunno really - I stopped as soon as I found the implicit assumption about 12.5% on the paths that require re-compression.  If I pick it up again it'll probably be after compressed ARC is mandatory!

akhamar commented:
Just saying thanks to @gmelikov ! Waiting for this nice feature 👍

ahrens commented:
@gmelikov We are happy to merge this PR as-is (after rebasing), if the only known "problem" is that L2ARC may not be entirely persistent across upgrade/downgrade from these bits (as described in https://github.com/openzfs/zfs/pull/9416/files#r422401476).

But I'm not sure if I understood your comment https://github.com/openzfs/zfs/pull/9416#issuecomment-757480204, are you saying that there is a problem with embedded block compression with this PR?  Are there any other known problems?

gmelikov commented:
@ahrens rebased and reworked for embedded_data feature:
- changed d_len to cover both embedded_data and "at least one saved sector" cases
- zfs_tests are happy too

Ready to review and merge.

problame commented:
I've been reading up on this PR while following @gmelikov 's comment in the [compressor versioning RFC](https://github.com/openzfs/zfs/issues/12840#issuecomment-991875296).
I only want to address the backward-compatibility aspect.

Abstractly, this change sets precedence for an unversioned change in compressor function behavior.
This change is like "do we compress at all or use plain text".
Other compressor code changes in other PRs are like "do we compress but the output will look slightly different, but be readable by older decompressors" .
As far as I understand the situation, it really doesn't make a difference to the core problem.
Namely, that we break any existing implicit or explicit assumptions about compression being bit-by-bit deterministic forever.

It seems like in this case we're ready to accept breaking these assumptions if only the L2ARC is affected.
The question is: **is it a principled decision, or is it only tolerable in this specific case** because it's just the threshold that changes and hence existing blocks are, barring pathological cases, still going to recompress to the original bit pattern.

Anyways, if we're going to merge this change, and are willing to refrain from updating other compressors a little longer, we should use the opportunity to gather some data from the field:
- if the l2arc read fails, retry with the original threshold
- if that retry read succeeds, this was a case where this change caused harm
  - => increment a (new) kstat for this
  - => maybe use `zpool status` or whatever to put this right in front of user's faces, so we get bug reports in this issue tracker

adamdmoss commented:
I gave this PR a spin for fun.

It indeed appears to work as intended; on both barely-compressible and moderately-compressible trees of data, the on-disk size is consistently a few % smaller.

Performance characteristics are quite interesting though; with zstd zfs (de)compression the sys-time for reading data with the new threshold is measurably 5-25% higher than with the old threshold.  With lz4 the sys-time difference is barely-measurable in my brief testing (I'm also using https://github.com/openzfs/zfs/pull/12805 :) ).

I didn't measure comparative write speed or other decompressors.  The results are about what I'd expect, on reflection, though I sure wish it wasn't so expensive to decompress nearly-uncompressible data.  Probably comes down to the algorithm since this mostly turns into big bites of memmove-like straight copies in lz4.  Conversely I imagine this change could be somewhat horrible for gzip's performance (**but** gzip or any other compressor except lz4/zstd has very limited value in modern openzfs anyway).

adamdmoss commented:
p.s. as far as I could tell, I didn't casually run into any l2arc weirdness.  So far.

behlendorf commented:
> The question is: is it a principled decision, or is it only tolerable in this specific case because it's just the threshold that changes and hence existing blocks are, barring pathological cases, still going to recompress to the original bit pattern.

There is a subtle technical issue at the heart of this which makes these two cases slightly different.  It's not well documented outside the source, but both the compressed ARC and encryption code depend on the invariant that when re-compressing a buffer it is expected to be bit-for-bit identical to the version of that buffer already stored on disk.  Changing only the compression threshold preserves this behavior which is why it's tolerable.  Changing the compressor itself breaks this behavior which is why it's more problematic and why we need to be very careful about updating the compressor.  There are certainly ways to do it, but it's more involved than it would appear on casual inspection of the code.

problame commented:
All right.
Opinions on my proposal to retry with the original threshold to gather statistics?

gmelikov commented:
@problame for these code paths this patch guarantees that zio_compress_data() will run with zero threshold, so it shouldn't break.

https://github.com/openzfs/zfs/pull/9416/files#diff-c6a4a29cdb673eca558f36f9d2546950d9c7de89b6d06fc77a4bd8583efe5965L9296-R9298

https://github.com/openzfs/zfs/pull/9416/files#diff-c6a4a29cdb673eca558f36f9d2546950d9c7de89b6d06fc77a4bd8583efe5965L1784-R1786

Theoretical problem prerequisites:
- pool with compressed block (new threshold - less than 12.5%)
- import it with pre-PR code, which will try to recompress it with 12.5% threshold

So your proposal with stats should be in pre-PR code. Or did I miss something?

adamdmoss commented on module/zfs/zio_compress.c:
```diff
@@ -143,8 +143,18 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len,
+	 */
+	if (compress_threshold > 0) {
+		d_len = MAX(BPE_PAYLOAD_SIZE,
+		    s_len - (s_len % compress_threshold));
```
It looks like this would produce an undesirable d_len (albeit in the safe(?) 0..s_len range) if s_len > compress_threshold.  And I can't quickly see/intuit a reason why s_len > compress_threshold is always true; if it is, then please add a comment. :)

adamdmoss replied:
Wondering if this is more readable / provable:
```c
	/*
	 * Usable compression thresholds are:
	 * - less than BPE_PAYLOAD_SIZE (embedded_data feature)
	 * - at least one saved sector
	 */
	if (compress_threshold > 0) {
		if (s_len > compress_threshold)
			d_len = MAX(BPE_PAYLOAD_SIZE, s_len - compress_threshold);
		else
			d_len = BPE_PAYLOAD_SIZE;

	} else {
		/* Special case for reproducibility of ARC checks */
		d_len = s_len;
	}
```

gmelikov replied:
What do you mean by `undesirable`? Could you please give an example?

`s_len - (s_len % compress_threshold)` will give d_len rounded to sector size (because compress_threshold is a max sector size if this incarnation of PR).

And I don't like additional `if`, looks redundant for me.

adamdmoss commented on module/zfs/zio_compress.c:
```diff
@@ -143,8 +143,18 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len,
+		/* Special case for reproducibility of ARC checks */
+		d_len = s_len;
+	}
 
```
I think we should `return (s_len);` right here if `d_len >= s_len` (which might now be the case if, say, `BPE_PAYLOAD_SIZE >= s_len` because although it won't waste any (more) physical space than avoiding compression, it's going to be a waste of CPU to go ahead with compression.

adamdmoss replied:
(or, uh, just check that s_len < BPE_PAYLOAD_SIZE back in the `if (compress_threshold > 0) {` path...)

adamdmoss replied:
ignore, brainfart

nabijaczleweli commented on man/man7/zfsprops.7:
```diff
@@ -900,13 +900,16 @@ zeroes (the NUL byte).
-ratio; for example, 8kB blocks on disks with 4kB disk sectors must compress to 1/2
-or less of their original size.
+All blocks are allocated as a whole number of sectors
+(chunks of 2^ashift bytes, e.g. 512B or 4KB).
```
```
.Pq 2^ Ns Sy ashift No bytes , e.g . Sy 512B No or Sy 4KB .
```

nabijaczleweli commented on man/man7/zfsprops.7:
```diff
@@ -900,13 +900,16 @@ zeroes (the NUL byte).
+If compression saves less than one whole sector,
+the block will be stored uncompressed.
+Therefore, blocks whose logical size is a small number of sectors will
+experience less compression ratio
```
"a lesser compression ratio" or "less compression"?

gmelikov replied:
I'm not a native speaker, but `less compression ratio` looks good for me. `less compression` == `less compression ratio` for me.

nabijaczleweli replied:
You can't have "less ratio", because ratio is in this case a property, which you either have or you don't (you could have "fewer ratios", but that'd mean something different). A ratio, by definition, can get bigger or smaller – so either "a smaller/lesser/&c. compression ratio" or just "less compression"; I'm more inclined toward just "less compression" because it's more obvious what that actually means.

gmelikov replied:
I didn't get everything, unfortunately, but I'm ok with "less compression", updated.

nabijaczleweli commented on man/man7/zfsprops.7:
```diff
@@ -900,13 +900,16 @@ zeroes (the NUL byte).
+the block will be stored uncompressed.
+Therefore, blocks whose logical size is a small number of sectors will
+experience less compression ratio
+(e.g. for recordsize=16K with 4K sectors, which have 4 sectors per block,
```
```
(e.g. for
.Sy recordsize Ns = Ns Sy 16K
with 4K sectors, ...
```

gmelikov replied:
Thank you, used your examples as a base.

gmelikov commented on include/sys/zio_compress.h:
```diff
@@ -184,7 +185,7 @@ extern int lz4_decompress_zfs(void *src, void *dst, size_t s_len, size_t d_len,
  */
 extern size_t zio_compress_data(enum zio_compress c, abd_t *src, void *dst,
-    size_t s_len, uint8_t level);
+    size_t s_len, uint8_t level, int compress_threshold);
```
I've thought about `saving_threshold`, but I still prefer `compress_threshold` more, comment two lines earlier should help.

gmelikov commented on module/zfs/arc.c:
```diff
@@ -1782,7 +1783,7 @@ arc_hdr_authenticate(arc_buf_hdr_t *hdr, spa_t *spa, uint64_t dsobj)
 		abd_take_ownership_of_buf(abd, B_TRUE);
 		csize = zio_compress_data(HDR_GET_COMPRESS(hdr),
-		    hdr->b_l1hdr.b_pabd, tmpbuf, lsize, hdr->b_complevel);
+		    hdr->b_l1hdr.b_pabd, tmpbuf, lsize, hdr->b_complevel, 0);
```
Maybe I'm wrong, but it's int and I prefer straightforward int here.

nabijaczleweli commented on man/man7/zfsprops.7:
```diff
@@ -896,13 +896,20 @@ zeroes (the NUL byte).
+.Sy recordsize Ns = Ns Sy 16K
+with
+.Sy 4K
+sectors, hich have 4 sectors per block,
```
typo: which

gmelikov replied:
Oops, thanks!

gmelikov commented:
Rebased.

gmelikov commented:
Gentle ping @behlendorf @ahrens , maybe @rincebrain would be interested in review too due to compression-related patches.

rincebrain commented:
Sure, why not. I'm rarely lacking for opinions.

I have no concerns about the code itself, as far as I can see, it does what it's supposed to, fairly simply.

But I had a couple thoughts, reading this.
* If I'm reading this right, the consensus is that this code explicitly avoids breaking on differing compression thresholds, L2ARC, and uncompressed ARC, and the older code avoids breaking on that by chance because it doesn't check "is the compression worth it" threshold anywhere in those codepaths? So we neatly sidestep being burned by that in this change, right?
* I'm somewhat worried about https://github.com/openzfs/zfs/pull/9416#issuecomment-995434622 - I'm going to try giving it a spin after my testbeds are finished trying to break #13078, but markedly changing decompression performance with no opt-out seems like a very serious problem, to the point that I'd revert this in my local tree if true. If I ruled the world, I'd probably want that as a dataset property, since you might want your ~never decompressed backups to compress as much as possible and your more live data to be cheaper, but not lz4 cheap; somehow, though, that's not the case, so all I can do is voice concern.
* Mostly, my thinking on the last point is that "someone might have picked a markedly different setting if this change were present when they wrote the data" - which could be good or bad, but the fact that it's true should give us a little pause. Sometimes that's fine (e.g. if SHA256 would become ~free on a lot of hardware tomorrow because we land SHA-NI/ARMv8 SHA2 support), but the prospect of hearing "I wrote the data with OpenZFS pre-3.0, so it reads faster" concerns me a lot.

rincebrain commented:
A concerning data point out of a larger set of experiments still running:
* On my Ryzen 5900X, a ~42 GB 128k dataset with about a 1.04x compressratio on lz4 stock and 1.05x compressratio on zstd-3/6 stock took an average of {28.772, 30.552, 30.012} seconds to read entirely off disk compressed pre-PR (using LZ4, zstd-3, and zstd-6, and zfs send without -c, dropping caches entirely between runs) and post-PR takes an average of {32.142, 32.456, 37.258} seconds. So that's, what, {11%,6%,24%} slower decompression for {.05%,.04%,.06%} savings? (Not a typo - 19.3 MB savings on 42 GB for LZ4.)

On a higher-compressratio dataset, as you might expect, the numbers are closer to noise (since we're already running the decompressor much of the time), but "10% slower for no savings" is worrying to me, and to the advice that using LZ4 even on relatively incompressible datasets is ~free.

gmelikov commented:
@rincebrain thank you, it's an interesting example of at least corner case!

Could you give some info about bench method? What is under pool? RAM or disk? Data source?

What I don't understand is - `19.3 MB savings` with (I think you have) 4K sector size is 19.3*1024/4 ~= 5000 saved blocks, worst case is 5000 of different 128K blocks ~=650MB of newly compressed blocks. For LZ4 32.142-28.772sec=3.37sec, 650MB/3.37s= 190MB/s decompression speed for 128k blocks, not so fast for LZ4. But for 512B sector it would be 1500MB/s and ~40000 newly-compressed blocks with total size of 5.05GB, looks more real, was it so? If it is, 512B sector with little gain is a really-really bad case here. Really interested to look at that data source!

So first thought is - compression gains UP to 12.5%-sector_size may constantly cost us `{11%,6%,24%}` of decompression CPU time in your example (disk IO speed aside). And in worst case we may pay this CPU cycles for diminishing space wins. But if we remember about disk IO,  faster the disk - less help of any compression.

Manual threshold override may help to maintain same decompression time as before (but not a silver bullet). Set larger threshold by default for 512B sector?

Note, that with ashift=12 (4K sector) only recordsizes more than 32K will be affected by 12.5% threshold removal (but 4K block for 512B sector). With usual 4K sector small recordsize IO (such as usual DB tune) should not be affected.

rincebrain commented:
It's ashift 12, single disk vdev, backed by an NVMe SSD, which averaged
28.414 seconds doing the same exercise on a copy of the dataset with
compression=off.

This dataset was my collection of LSI^WAvago^WBroadcom firmware updates,
which is mostly ZIP files and occasionally unpacked contents of those ZIP
files, just because it was around the right size and compression ratio for
what I wanted.

- Rich

On Wed, Mar 2, 2022 at 10:07 AM George Melikov ***@***.***>
wrote:

> @rincebrain <https://github.com/rincebrain> thank you, it's an
> interesting example of at least corner case!
>
> Could you give some info about bench method? What is under pool? RAM or
> disk? Data source?
>
> What I don't understand is - 19.3 MB savings with (I think you have) 4K
> sector size is 19.3*1024/4 ~= 5000 saved blocks, worst case is 5000 of
> different 128K blocks ~=650MB of newly compressed blocks. For LZ4
> 32.142-28.772sec=3.37sec, 650MB/3.37s= 190MB/s decompression speed for 128k
> blocks, not so fast for LZ4. But for 512B sector it would be 1500MB/s and
> ~40000 newly-compressed blocks with total size of 5.05GB, looks more real,
> was it so? If it is, 512B sector with little gain is a really-really bad
> case here. Really interested to look at that data source!
>
> So first thought is - compression gains UP to 12.5%-sector_size may
> constantly cost us {11%,6%,24%} of decompression CPU time in your example
> (disk IO speed aside). And in worst case we may pay this CPU cycles for
> diminishing space wins. But if we remember about disk IO, faster the disk -
> less help of any compression.
>
> Manual threshold override may help to maintain same decompression time as
> before (but not a silver bullet). Set larger threshold by default for 512B
> sector?
>
> Note, that with ashift=12 (4K sector) only recordsizes more than 32K will
> be affected by 12.5% threshold removal (but 4K block for 512B sector). With
> usual 4K sector small recordsize IO (such as usual DB tune) should not be
> affected.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openzfs/zfs/pull/9416#issuecomment-1057036280>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUI7LSWTWPZSCM3RDCNY3U557Z7ANCNFSM4I5ZBQDA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gmelikov commented:
> It's ashift 12

Then I'm bit concerned about decompression speed, it's too slow. Sorry to be noisy, but did you set ashift on pool creation? Most on NVMEs like to pretend as 512bytes-sector-size. Just to be sure.

rincebrain commented:
Yes, I set ashift explicitly at creation.

On Wed, Mar 2, 2022, 12:19 PM George Melikov ***@***.***>
wrote:

> It's ashift 12
>
> Then I'm bit concerned about decompression speed, it's too slow. Sorry to
> be noisy, but did you set ashift on pool creation? Most on NVMEs like to
> pretend as 512bytes-sector-size. Just to be sure.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openzfs/zfs/pull/9416#issuecomment-1057180093>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUI7LB4US3MCNOK3SHDALU56PIPANCNFSM4I5ZBQDA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

adamdmoss commented:
aside: now I have a brainworm who is wondering if moving the existing 12.5% threshold in the *other* direction could be an interesting perf win, especially for those like me who kinda treat lz4 as a throughput and ARC residency optimization (for the higher-compression cases) rather than a pure space saving.

gmelikov commented:
What's interested, with 128K recordsize sector_size threshold is ~3%, not 0.*%.

<body>


recordsize |  minimal compression threshold with 4K sector size (% of recordsize)
-- | --
4096 | 100
8192 | 50
16384 | 25
32768 | 12.5
65536 | 6.25
131072 | 3.125
262144 | 1.5625
524288 | 0.78125
1048576 | 0.390625
2097152 | 0.1953125
4194304 | 0.09765625
8388608 | 0.048828125
16777216 | 0.0244140625


</body>

And I can't make LZ4 decompression speed for 128K block size worse than 600MB/s on brd device, even random (yet).

Meanwhile, I'll try to look at "threshold per dataset", but it may not be fast (heh, this PR is nearly 2.5 years old).

rincebrain commented:
I can also send you the send stream(s) for the low-compressibility dataset
I'm using off-bug, if you'd like. (Off-bug just because I don't really want
to post a 20+ GB link somewhere that'll be publicly spidered.)

(I was anticipating the effect to be worse at higher recordsizes, so I was
going to try that next, as well. I've got a short spreadsheet that's not
completely filled out yet of experiments and results.)

I also might take a crack at a property for it, if you're not opposed but
just don't necessarily have the bandwidth in the near term.

On Wed, Mar 2, 2022 at 2:36 PM George Melikov ***@***.***>
wrote:

> What's interested, with 128K recordsize sector_size threshold is ~3%, not
> 0.*%.
> recordsize minimal compression threshold with 4K sector size (% of
> recordsize)
> 4096 100
> 8192 50
> 16384 25
> 32768 12.5
> 65536 6.25
> 131072 3.125
> 262144 1.5625
> 524288 0.78125
> 1048576 0.390625
> 2097152 0.1953125
> 4194304 0.09765625
> 8388608 0.048828125
> 16777216 0.0244140625
>
> And I can't make LZ4 decompression speed for 128K block size worse than
> 600MB/s on brd device, even random (yet).
>
> Meanwhile, I'll try to look at "threshold per dataset", but it may not be
> fast (heh, this PR is nearly 2.5 years old).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openzfs/zfs/pull/9416#issuecomment-1057305012>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUI7OQY7K6BSKFVDKZJS3U567MTANCNFSM4I5ZBQDA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gmelikov commented:
> I also might take a crack at a property for it, if you're not opposed but
just don't necessarily have the bandwidth in the near term.

Not at all, it's a great idea, main thing is to get this property without new feature flag, thank you! Will be glad to help with whatever I can and review it. Feel free to use this PR as a basis.

gmelikov commented:
I've rebased it and change something:
- I've got back 12.5% threshold, let's deal with it in another PR

, so we can merge it and we shouldn't have visible performance degradation.

Why we may need this PR:
- we'll have expected behavior on compressions (especially encryption recompression cases) even if we change thresholds in future, the earlier we can have this logic in code the better for future us
- now we'll have better thresholds, and at least LZ4 and ZSTD can fail back faster if they can not get expected compression gains

Failed tests don't look for me as correlated with this PR, I can see same fails on other fresh PRs.

rincebrain commented on module/zfs/dmu_recv.c:
```diff
@@ -1408,7 +1408,7 @@ do_corrective_recv(struct receive_writer_arg *rwa, struct drr_write *drrw,
 		uint64_t csize = zio_compress_data(BP_GET_COMPRESS(bp),
 		    abd, &buf, abd_get_size(abd),
-		    rwa->os->os_complevel);
+		    rwa->os->os_complevel, 0);
```
won't this break corrective receive in the same way as e.g. triggering gang blocks when they didn't exist before does in the test suite, by emitting a compressed record where one might not have existed before?

gmelikov replied:
`BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF` in condition checks that on-disk BP has compressed flag, so the block was successfully compressed before (even with previous thresholds). And we already don't look at `csize` here, so we'll already write it as is. Hope I didn't miss anything.

gmelikov replied:
@alek-p I'd be glad if you may have time to recheck me here as the author of corrective recv, thank you!

rincebrain replied:
I think I agree with you, I just am wary because I know there's already existing failure modes around corrective receive if the output is not JUUUST right, but reading more context, I don't see a way this could be wrong unless the compressor fails nondeterministically, which would break many things already.

gmelikov replied:
> unless the compressor fails nondeterministically, which would break many things already.

you reminded me that compressor code has fast fail back if it can't allocate memory, and this case may give us problems everywhere where we wait for idempotent compression results. But it's not a new problem, so just a thought for now.

alek-p replied:
I _think_ there might be corner cases that could impact corrective recv functionality if we have an old system without this patch that is using a compressed/raw sendfile from a newer system with this patch for corrective recv.
I believe the workaround would be to generate a sendfile that is not compressed (not using "zfs send -c/-w")

rincebrain replied:
Older hosts I think would only break on -w, and you can't really do much about that. -c you could just not use -c or use zstream decompress as an input and make it redo it.

gmelikov commented:
@behlendorf @ahrens it's ready to review, please see https://github.com/openzfs/zfs/pull/9416#issuecomment-1962330982

amotin commented on module/zfs/zio_compress.c:
```diff
@@ -145,8 +144,19 @@ zio_compress_data(enum zio_compress c, abd_t *src, void **dst, size_t s_len,
+	 */
+	if (compress_threshold > 0) {
+		d_len = MAX(BPE_PAYLOAD_SIZE,
+		    s_len - MAX(s_len % compress_threshold, s_len >> 3));
```
I think `s_len >> 3` here could also be rounded up or down to `compress_threshold`, otherwise it may produce half-full last sector which either give us no real saving or will be excessively strict when it is not required.

gmelikov replied:
I propose to leave `s_len >> 3` as is in this exact patch to make it equal with previous calculations, to minimize threshold changes here.

But it's a great addition for next patches about default threshold changes!

amotin replied:
Not that I strongly object, but IMO it is not exactly a rocket science to postpone.  I do see that this PR is hanging here for years, but I would prefer it to be done once and done right.  What would you say about this:
```suggestion
                d_len = s_len - (s_len >> 3);
		d_len = MAX(BPE_PAYLOAD_SIZE, d_len - d_len % compress_threshold);
```
On a downside this still does not account for spa_min_alloc, which we do not have here now.  In case lsize/s_len minus 12.5% and rounded down to spa_gcd_alloc is less then spa_min_alloc, we should not even try to compress beyond BPE_PAYLOAD_SIZE, since it will not have sense.  I am thinking if we could move all this math out of here and may be into a separate function alike to zio_roundup_alloc_size() and pass to zio_compress_data() as fully prepared value.

gmelikov replied:
@amotin thank you for detailed comment, I moved it into separate logic, now it looks much better indeed! I've used your math, agree with it.

amotin commented on module/zfs/arc.c:
```diff
@@ -10442,7 +10445,8 @@ l2arc_log_blk_commit(l2arc_dev_t *dev, zio_t *pio, l2arc_write_callback_t *cb)
 	psize = zio_compress_data(ZIO_COMPRESS_LZ4,
-	    abd_buf->abd, (void **) &tmpbuf, sizeof (*lb), 0);
+	    abd_buf->abd, (void **) &tmpbuf, sizeof (*lb), 0,
+	    1 << dev->l2ad_vdev->vdev_ashift);
```
This also needs to be updated to the zio_get_compression_max_size() logic, just with the 1<<ashift instead of spa values.

amotin commented on module/zfs/zio_compress.c:
```diff
@@ -126,9 +122,9 @@ zio_compress_zeroed_cb(void *data, size_t len, void *private)
 
 size_t
 zio_compress_data(enum zio_compress c, abd_t *src, void **dst, size_t s_len,
-    uint8_t level)
+    uint8_t level, size_t max_size)
```
The list of arguments looks messy to me.  I would call it d_len and put after s_len.

gmelikov replied:
Great to have fresh eyes on it, d_len makes it consistent with other *(de)compress_data() functions, done.

amotin commented on module/zfs/arc.c:
```diff
@@ -10442,7 +10445,9 @@ l2arc_log_blk_commit(l2arc_dev_t *dev, zio_t *pio, l2arc_write_callback_t *cb)
 	psize = zio_compress_data(ZIO_COMPRESS_LZ4,
-	    abd_buf->abd, (void **) &tmpbuf, sizeof (*lb), 0);
+	    abd_buf->abd, (void **) &tmpbuf, sizeof (*lb),
+	    sizeof (*lb) - sizeof (*lb) % (1 << dev->l2ad_vdev->vdev_ashift),
```
With `sizeof (*lb)` always being 64KB the result will always be `sizeof (*lb)`.  Not enough.

amotin reviewed (approved):
I still don't very like the L2ARC chunk. It is a custom code, plus with sizeof (*lb) always being 64KB the reminder will always be zero.  But whatever...

behlendorf commented on module/zfs/zio.c:
```diff
@@ -1674,6 +1675,20 @@ zio_roundup_alloc_size(spa_t *spa, uint64_t size)
+{
+	size_t d_len;
+
+	// minimun 12.5% should be saved (legacy value, may be changed later)
```
```suggestion
	/* minimum 12.5% should be saved (legacy value, may be changed later) */
```

amotin commented:
> agree with you, but I don't have better thoughts for now.

@gmelikov I would make zio_get_compression_max_size() global and passed to it instead of `spa` exactly `spa->spa_gcd_alloc` and `spa->spa_min_alloc` in normal case, and `1 << dev->l2ad_vdev->vdev_ashift` twice in case of L2ARC.

behlendorf commented:
@gmelikov can you rebase this again to get a fresh run.  We shouldn't be seeing all of those test builders failing.

behlendorf commented:
@gmelikov thanks, I haven't dug in to this but it looks like tho `bclone` ZTS failures where somehow introduced by this PR.  Those tests reliably pass in the master branch.

gmelikov commented:
@behlendorf sorry that I didn't post it earlier, I've already found the root cause https://github.com/openzfs/zfs/issues/16237

It would be best to rework zle_compress somehow, but there's a great risk to miss some cases, so for a start I've added additional bytes to d_len for BPE_PAYLOAD_SIZE case https://github.com/openzfs/zfs/pull/9416/commits/17efc68ef24f03463a605ddac8c011d22c35d227 to run tests with CI.

rincebrain commented:
I'd just special-case ZLE for the check, tbh - it was introduced for DDT records where it ignores the threshold requirements anyway, and anyone who's using it is already in an unfortunate place.

More broadly, I'd replace user-visible zle with a new compression option and keep actual zle for older records and for the DDT using it, though I think all the "better" zle implementations are not really license-friendly if we wanted to do anything but improve this case for that.

gmelikov commented:
Yep, I've disabled any new d_len variant for ZLE explicitly for now in https://github.com/openzfs/zfs/pull/9416/commits/e0ba8b45ae1ef48aded7025480b8865438346a9a , CI says that bclone tests are ok now, rebase again just in case.

gmelikov commented:
@amotin thank you, applied as is, I've put arc_hdr_authenticate changes in a separate commit to add some context for this change.

gmelikov commented on module/zfs/arc.c:
```diff
@@ -1786,9 +1787,13 @@ arc_hdr_authenticate(arc_buf_hdr_t *hdr, spa_t *spa, uint64_t dsobj)
+		    hdr->b_complevel);
+		if (csize >= lsize || csize > psize) {
+				ret = SET_ERROR(EIO);
+				goto error;
```
Thank you, missed it too, fixed.

rincebrain commented:
Is there any reason not to introduce a tunable with this PR to let you change it?

Otherwise it seems a little strange to have the PR just do all the plumbing and not expose it, when that seems like a minor addition, and was the original goal, I think...

gmelikov commented:
@rincebrain It would be great to merge it as soon as possible (and maybe even backport it into stable release) to have code that can withstand different thresholds in future, no matter how this would be used.

After that I thought about exposing it as a dataset property. Plus I don't want to make all-in-one PRs to make the process easier. This is my most long-standing PR as for now, nearly 5 years, wow :)

rincebrain commented:
Avoiding all-in-one PRs because they're hard to get reviewed only makes sense as an argument in projects that review smaller PRs any faster, IMO.

Oh well, I've been running my own implementation of this as a dataset property out of tree for a while now, I guess I'll just keep maintaining it.

jumbi77 commented:
@behlendorf Can you may give @gmelikov a hint whats left/missing to get this integrated? And of course thanks to all participants.

gmelikov commented:
@behlendorf @tonyhutter freshly rebased, it's ready to merge from my side.

jumbi77 commented:
Politely ping @tonyhutter for review/merging.

tonyhutter reviewed (approved):
Sorry this fell off our radar - approved

gmelikov commented:
Don't see CI fails related to changes, previous runs were successfull, rebased up to date.

behlendorf commented:
@gmelikov the Fedora build failures do look like they're introduced by this PR.

gmelikov commented:
@behlendorf not sure what was with fedora runner, after rebase I've got 2 green runs.

But I had flaky tests (they've failed only once):
- import_devices_missing
- redundancy_draid_spare1 (saw in some other PRs too)

behlendorf reviewed (approved):
Thanks for updating this.  Looks good, the CI failures were unrelated.
